### PR TITLE
add Category model

### PR DESCRIPTION
**Pull request by qreqit**

### CODE REVIEW

File: src/main/java/mate/academy/springbootwebgreqit/dto/category/CreateCategoryRequestDto.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package mate.academy.springbootwebgreqit.dto.category;
+
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class CreateCategoryRequestDto {
+    private String name;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/dto/category/CreateCategoryRequestDto.java:9]: Should we validate this DTO?

File: src/main/java/mate/academy/springbootwebgreqit/controller/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        return categoryService.update(id, categoryDto);
+    }
+
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @DeleteMapping("/{id}")
+    public void deleteCategory(@PathVariable Long id){
+        categoryService.deleteById(id);
+    }
+
+    @GetMapping
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/controller/CategoryController.java:47]: Is this `@GetMapping("/{id}/books")` endpoint?

File: src/main/java/mate/academy/springbootwebgreqit/controller/CategoryController.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+public class CategoryController {
+    private final CategoryService categoryService;
+
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @PostMapping
+    public CategoryDto createCategory(@Valid @RequestBody CategoryDto categoryDto){
+        return categoryService.save(categoryDto);
+    }
+
+    @GetMapping
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/controller/CategoryController.java:25]: Should we add `@PreAuthorize` or you have other opinion?

File: src/main/java/mate/academy/springbootwebgreqit/controller/CategoryController.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    public List<CategoryDto> getAll(){
+        return categoryService.findAll();
+    }
+
+    @GetMapping("/{id}")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/controller/CategoryController.java:30]: same

File: src/main/java/mate/academy/springbootwebgreqit/model/Book.java [hunk 5]
```diff
@@ -31,5 +35,14 @@ public class Book {
     private String description;
     @Column(nullable = false, length = 255)
     private String coverImage;
+    @ManyToMany(fetch = FetchType.LAZY)
+    @JoinTable(
+            name = "book_category",
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/model/Book.java:40]:
```suggestion
            name = "books_categories",
```

File: src/main/java/mate/academy/springbootwebgreqit/model/Category.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false)
+    private String name;
+    private String description;
+    @ManyToMany(mappedBy = "categories", fetch = FetchType.LAZY)
+    @ToStringExclude
+    @EqualsAndHashCode.Exclude
+    private Set<Book> book = new HashSet<>();
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/model/Category.java:26]:
```suggestion
```

File: src/main/java/mate/academy/springbootwebgreqit/model/Book.java [hunk 7]
```diff
@@ -31,5 +35,14 @@ public class Book {
     private String description;
     @Column(nullable = false, length = 255)
     private String coverImage;
+    @ManyToMany(fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/model/Book.java:38]:
```suggestion
    @ManyToMany
```

File: src/main/java/mate/academy/springbootwebgreqit/service/impl/CategoryServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    @Override
+    public List<CategoryDto> findAll() {
+        return categoryRepository.findAll().stream()
+                .map(categoryMapper::toDto)
+                .toList();
+    }
+
+    @Override
+    public CategoryDto getById(Long id) {
+        Category category = categoryRepository.findById(id).orElseThrow(() -> new  EntityNotFoundException("Can't find by id" + id));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/service/impl/CategoryServiceImpl.java:36]:
```suggestion
        Category category = categoryRepository.findById(id)
            .orElseThrow(() -> new  EntityNotFoundException("Can't find by id" + id));
```

not clear message

File: src/main/java/mate/academy/springbootwebgreqit/service/impl/CategoryServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        return categoryMapper.toDto(category);
+    }
+
+    @Override
+    public CategoryDto save(CategoryDto categoryDto) {
+        Category category = categoryMapper.toEntity(categoryDto);
+        Category savedCategory = categoryRepository.save(category);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/service/impl/CategoryServiceImpl.java:43]:
```suggestion
        categoryRepository.save(category);
```

File: src/main/java/mate/academy/springbootwebgreqit/service/impl/CategoryServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        return categoryMapper.toDto(savedCategory);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/service/impl/CategoryServiceImpl.java:44]:
```suggestion
        return categoryMapper.toDto(category);
```

File: src/main/java/mate/academy/springbootwebgreqit/service/impl/CategoryServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    @Override
+    public CategoryDto update(Long id, CategoryDto categoryDto) {
+        Category category = categoryRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Can't find category by id" + id));
+        Category updatedCategory = categoryRepository.save(category);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/service/impl/CategoryServiceImpl.java:51]:
```suggestion
        mapper.updateCategoryFromDto(categoryDto, category);
        return categoryMapper.toDto(categoryRepository.save(category));
```

File: src/main/java/mate/academy/springbootwebgreqit/service/impl/CategoryServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    public void deleteById(Long id) {
+        categoryRepository.deleteById(id);
+    }
+
+    @Override
+    public List<BookDtoWithoutCategotyIds> findBooksByCategoryId(Long categotyId) {
+        List<Book> books = bookRepository.findByCategoryId(categotyId);
+        return books.stream()
+                .map(bookMapper::toDtoWithoutCategories)
+                .collect(Collectors.toList());
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/service/impl/CategoryServiceImpl.java:65]:
```suggestion
                .toList();
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Please, add changelogs for categories table and books-categories

File: src/main/java/mate/academy/springbootwebgreqit/dto/category/CreateCategoryRequestDto.java (hunk 1 above)
Comment (qreqit) [reply src/main/java/mate/academy/springbootwebgreqit/dto/category/CreateCategoryRequestDto.java:9]: yes

File: src/main/java/mate/academy/springbootwebgreqit/controller/CategoryController.java (hunk 3 above)
Comment (qreqit) [reply src/main/java/mate/academy/springbootwebgreqit/controller/CategoryController.java:25]: yeah, we can add @PreAuthorize with User role

Comment (qreqit) [reply src/main/java/mate/academy/springbootwebgreqit/controller/CategoryController.java:25]: hmmm, i think it's ok when user can find all books

File: src/main/java/mate/academy/springbootwebgreqit/controller/CategoryController.java (hunk 2 above)
Comment (qreqit) [reply src/main/java/mate/academy/springbootwebgreqit/controller/CategoryController.java:47]: yeah

File: src/main/java/mate/academy/springbootwebgreqit/controller/CategoryController.java (hunk 3 above)
Comment (qreqit) [reply src/main/java/mate/academy/springbootwebgreqit/controller/CategoryController.java:25]: yes, we should add `@PreAuthorize`

File: src/main/java/mate/academy/springbootwebgreqit/model/Book.java [hunk 13]
```diff
@@ -31,5 +35,14 @@ public class Book {
     private String description;
     @Column(nullable = false, length = 255)
     private String coverImage;
+    @ManyToMany
+    @JoinTable(
+            name = "book_categories",
```
Comment (sarakhmen) [inline src/main/java/mate/academy/springbootwebgreqit/model/Book.java:40]:
```suggestion
            name = "books_categories",
```

**Review by sarakhmen — APPROVED** [review]
lgtm!

**Pull request merged**